### PR TITLE
Task/864 create crop raster

### PR DESCRIPTION
**Pull request by rasrocha**
Please review my PR. This includes the crop_raster function and respective test. The tests should probably be more complete, but for now should be enough.

### CODE REVIEW

File: src/coresyf_image_crop/coresyf_image_crop.py [hunk 1]
```diff
@@ -76,6 +77,40 @@ def get_shapefile_crs(data_source):
     return epsg_code
 
 
+def crop_raster(input_raster, output_raster_file, polygon_extent, dest_crs):
+    '''
+    Crops the image specified by input_raster using the limits defined in
+    polygon_extent (buffer already applied).
```
Comment (rmcapote) [inline src/coresyf_image_crop/coresyf_image_crop.py]: Remove "(buffer already applied)". It is not important for the function description if the polygon has buffer already applied or not. It is just a suggestion that helps to keep the function description generic.

File: src/coresyf_image_crop/coresyf_image_crop.py [hunk 2]
```diff
@@ -76,6 +77,40 @@ def get_shapefile_crs(data_source):
...
+
+    input_raster: path to the image to be cropped.
+    output_raster_file: the output file path.
```
Comment (rmcapote) [inline src/coresyf_image_crop/coresyf_image_crop.py]: Both variables are paths, so we can try to give them similar names such as: 
1. "input_raster_path" and "output_raster_path" or
2. "input_path" and "output_path"

File: src/coresyf_image_crop/coresyf_image_crop.py [hunk 3]
```diff
@@ -76,6 +77,40 @@ def get_shapefile_crs(data_source):
...
+    polygon_extent: POLYGON object to be used for the crop limits.
+    dest_crs: the CRS of the output raster.
```
Comment (rmcapote) [inline src/coresyf_image_crop/coresyf_image_crop.py]: Perhaps change "dest_crs" to "output_crs".

File: src/coresyf_image_crop/coresyf_image_crop.py [hunk 4]
```diff
@@ -76,6 +77,40 @@ def get_shapefile_crs(data_source):
...
+    '''
+    bounds = polygon_extent.GetEnvelope()
+    command_opts = str(bounds[0]) + ' ' + str(bounds[2]) + ' ' \
+        + str(bounds[1]) + ' ' + str(bounds[3])
+    gdal_command = 'gdalwarp' + ' -t_srs EPSG:' + str(dest_crs) + ' -te ' \
```
Comment (rmcapote) [inline src/coresyf_image_crop/coresyf_image_crop.py]: This looks more fancy like this:
```
    gdal_command = 'gdalwarp -t_srs EPSG:{} -te {} {} {}'.format(
                    dest_crs, command_opts, input_raster, output_raster_file)
```

File: src/coresyf_image_crop/coresyf_image_crop.py [hunk 5]
```diff
@@ -76,6 +77,40 @@ def get_shapefile_crs(data_source):
...
+                   + str(command_opts) + ' ' + input_raster \
+                   + ' ' + output_raster_file
+    try:
+        process = subprocess.Popen(gdal_command,
```
Comment (rmcapote) [inline src/coresyf_image_crop/coresyf_image_crop.py]: Perhaps we should see this with @91nunocosta. We are repeating code here to execute the command line.

File: src/coresyf_image_crop/tests/test_image_crop.py [hunk 6]
```diff
@@ -18,6 +19,8 @@ def setUp(self):
             "-54172.937265817548905 107225.478697661776096 0," \
             "-61575.57812574980926 107225.478697661776096 0,"\
             "-61575.57812574980926 103754.81329901740537 0))"
+
```
Comment (rmcapote) [inline src/coresyf_image_crop/tests/test_image_crop.py]: Perhaps remove this blank line here.

File: src/coresyf_image_crop/tests/test_image_crop.py [hunk 7]
```diff
@@ -29,5 +32,15 @@ def test_get_shapefile_polygon_extent(self):
...
     def test_get_shapefile_crs(self):
-        espg_code = get_shapefile_crs(self.data_source)
-        self.assertEqual(espg_code, 3763)
\ No newline at end of file
+        epsg_code = get_shapefile_crs(self.data_source)
+        self.assertEqual(epsg_code, self.epsg_code)
+
+    def test_crop_raster(self):
+        test_image = 'test_data/Aveiro_resampled.tif'
+        output_image = os.path.join(TEMP_PATH, 'output')
```
Comment (rmcapote) [inline src/coresyf_image_crop/tests/test_image_crop.py]: You can move this and previous line to the "setup" method.

**Review by rmcapote — CHANGES_REQUESTED** [review]
Please consider my comments.

File: src/coresyf_image_crop/coresyf_image_crop.py (hunk 1 above)
Comment (rasrocha) [reply src/coresyf_image_crop/coresyf_image_crop.py]: Done

File: src/coresyf_image_crop/coresyf_image_crop.py (hunk 2 above)
Comment (rasrocha) [reply src/coresyf_image_crop/coresyf_image_crop.py]: Done

File: src/coresyf_image_crop/coresyf_image_crop.py (hunk 3 above)
Comment (rasrocha) [reply src/coresyf_image_crop/coresyf_image_crop.py]: Done

File: src/coresyf_image_crop/coresyf_image_crop.py (hunk 4 above)
Comment (rasrocha) [reply src/coresyf_image_crop/coresyf_image_crop.py]: Done.

File: src/coresyf_image_crop/tests/test_image_crop.py (hunk 7 above)
Comment (rasrocha) [reply src/coresyf_image_crop/tests/test_image_crop.py]: done

File: src/coresyf_image_crop/tests/test_image_crop.py (hunk 6 above)
Comment (rasrocha) [reply src/coresyf_image_crop/tests/test_image_crop.py]: Due to moving two lines from test_crop_raster, the blank line now makes more sense to be here.

File: src/coresyf_image_crop/coresyf_image_crop.py [hunk 8]
```diff
@@ -78,9 +77,27 @@ def get_shapefile_crs(data_source):
...
+        dest_crs: the CRS of the output raster.
+        '''
+        envelope = polygon_extent.GetEnvelope()
+        bounds = "{} {} {} {}".format(str(envelope[0]), str(envelope[2]),
+                                      str(envelope[1]), str(envelope[3]))
+
+        command_template = "gdalwarp -t_srs EPSG:{} -te {}".format(
+            str(output_crs), bounds)
+        command_template += ' {Ssource} {Ttarget}'
+        self.invoke_shell_command(command_template, **self.bindings)
```
Comment (rmcapote) [inline src/coresyf_image_crop/coresyf_image_crop.py:91]: Perhaps, to be compliant with other tool's logs, add the following log:
`self.logger.debug('Command: %s', str(command_template))`

**Review by rmcapote — APPROVED** [review]

**Pull request merged**